kernel/src/column_mapping.rs

```diff
@@ -0,0 +1,57 @@
+//! Code to handle column mapping, including modes and schema transforms
+
+use crate::{
+    schema::{ColumnMetadataKey, MetadataValue, StructField},
+    DeltaResult, Error,
+};
+use serde::{Deserialize, Serialize};
+
+/// Modes of column mapping a table can be in
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub enum ColumnMappingMode {
+    None,
+    Id,
+    Name,
+}
+
+// key to look in metadata.configuration for to get column mapping mode
+pub(crate) const COLUMN_MAPPING_MODE_KEY: &str = "delta.columnMapping.mode";
+
+impl TryFrom<&str> for ColumnMappingMode {
+    type Error = Error;
+
+    fn try_from(mode: &str) -> DeltaResult<Self> {
+        match mode {
+            "none" => Ok(ColumnMappingMode::None),
+            "id" => Ok(ColumnMappingMode::Id),
+            "name" => Ok(ColumnMappingMode::Name),
+            _ => Err(Error::invalid_column_mapping_mode(mode)),
+        }
+    }
+}
+
+/// Transform a logical `StructField` into the field as it should be read from parquet. *Only* valid
+/// to call this from a scan operating in ColumnMappingMode::Name mode.
+///
+/// This returns both the field and the name of the field. The name has the same lifetime as the
+/// logical field
+pub(crate) fn get_name_mapped_physical_field(
+    logical_field: &StructField,
+) -> DeltaResult<(StructField, &str)> {
+    match logical_field.metadata.get(ColumnMetadataKey::ColumnMappingPhysicalName.as_ref()) {
+        Some(val) => match val {
+            MetadataValue::Number(_) => {
+                Err(Error::generic("{ColumnMetadataKey::ColumnMappingPhysicalName} must be a string in name mapping mode"))
+            }
+            MetadataValue::String(name) => {
+                Ok((
+                    StructField::new(name, logical_field.data_type().clone(), logical_field.is_nullable()),
+                    name
+                ))
+            }
+        }
+        None => {
+            Err(Error::generic("fields MUST have a {ColumnMetadataKey::ColumnMappingPhysicalName} key in their metadata in name mapping mode"))
+        }
+    }
+}
```

Comment:
in `delta-rs` we just extended the `StructField` with `get_physical_name`. Don't have too string feeling about this, since its only internal, but personally I find it a little bit more convenient to work with.

Reply:
"more convenient" as in `field.get_physical_name()` rather than `get_name_mapped_physical_field(field)`? I would tend to agree with that.

Reply:
Also, based on my experience debugging column mapping issues in delta-spark -- we should strongly consider making logical vs. physical be a first class concept everywhere in kernel. At any point when we have a field name (or schema), it should be immediately clear from the context whether that name/schema is logical or physical. In case column mapping is disabled, logical and physical name are the same. But almost all code shouldn't have to care about that corner case and should instead assume column mapping is enabled.

Otherwise, we risk trying to map a physical name, or failing to map a logical name.

Reply:
Basically, the column mapping function would take the current column mapping mode as input, and then would use that to require the appropriate metadata is available:
```rust
impl StructField {
  // Returns the physical name of this field, based on the column mapping mode in effect.
  fn physical_name(&self, mode: ColumnMappingMode): DeltaResult<&str> {
    let name = logical_field.metadata.get(ColumnMetadataKey::ColumnMappingPhysicalName.as_ref());
    match (mode, name) {
      (ColumnMappingMode::None, None) => Ok(self.name.as_str()),
      (ColumnMappingMode::Name, Some(MetadataValue::String(name))) => Ok(name.as_str()),
      _ => Err(Error::generic(...)),
      
    }
  }
}
```

Reply:
Yep, I've done mostly that. 

Re the logical/physical naming comment, makes a lot of sense, and I've converted a lot of stuff to be explicit here. I'll take another pass to see what else could be renamed though.